// @aws-cdk/aws-lambda-event-sources/test/integ.kinesiswithdlq.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const kinesis = require("@aws-cdk/aws-kinesis");
const lambda = require("@aws-cdk/aws-lambda");
const sqs = require("@aws-cdk/aws-sqs");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
/*
 * Stack verification steps:
 * * aws kinesis put-record --stream-name <value of stack output: InputKinesisStreamName> --partition-key 123 --data testdata
 * * aws sqs receive-message --queue-url <value of stack output: DlqSqsQueueUrl> --max-number-of-messages 1 --query 'Messages[0].Body'
 * The last command should return a string that contains the Lambda function ARN in it.
 */
async function handler(event) {
    // tslint:disable-next-line:no-console
    console.log('event:', JSON.stringify(event, undefined, 2));
    throw new Error();
}
class KinesisWithDLQTest extends core_1.Stack {
    constructor(scope, id) {
        super(scope, id);
        const fn = new lambda.Function(this, 'F', {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.fromInline(`exports.handler = ${handler.toString()}`),
        });
        new core_1.CfnOutput(this, 'FunctionArn', { value: fn.functionArn });
        const stream = new kinesis.Stream(this, 'S');
        new core_1.CfnOutput(this, 'InputKinesisStreamName', { value: stream.streamName });
        const dlq = new sqs.Queue(this, 'Q');
        new core_1.CfnOutput(this, 'DlqSqsQueueUrl', { value: dlq.queueUrl });
        fn.addEventSource(new lib_1.KinesisEventSource(stream, {
            startingPosition: lambda.StartingPosition.TRIM_HORIZON,
            onFailure: new lib_1.SqsDlq(dlq),
            retryAttempts: 0,
        }));
    }
}
const app = new core_1.App();
new KinesisWithDLQTest(app, 'lambda-event-source-kinesis-with-dlq');
app.synth();
